### ISSUE TITLE
TagHelper: <input>'s asp-for is binding to query-string value instead of view-model property value

### ISSUE THREAD

**Issue by mcrsw** [aspnet/Mvc#5525]
Microsoft.AspNetCore.Mvc:  version 1.0.1

Steps to Reproduce:
1. Open VS 2015
2. Select New | Project
3. Select "ASP.NET Core Web Application (.NET Core)"
4. Enter a (project) Name, e.g. "AspNetCoreTagBug"
5. Choose an appropriate Location
6. Click "OK"
7. Select "Web Application"
8. Click "OK"
9. Open HomeController.cs, and add the following code to the end of the class:
```
    public IActionResult ShowBug(string name)
    {
        var vm = new ShowBugViewModel
        {
            Name = "Apple",
        };

        return View(vm);    // See the new view: Home/ShowBug.cshtml
    }
```
10. Paste the following class definition into the HomeController.cs file (below the closing } of the HomeController class definition)
```
    public class ShowBugViewModel
    {
        public string Name { get; set; }
    }
```
11. Create a new MVC view under Views | Home, called ShowBug.cshtml
12. Paste the following code into the ShowBug.cshtml file:
```
    @using AspNetCoreTagBug.Controllers;
    @model AspNetCoreTagBug.Controllers.ShowBugViewModel

    <h1>Show Me The Bug</h1>

    <a href="~/Home/ShowBug?Name=orange">Show me the bug (GET with query string parameters)</a>

    <hr />

    <a href="~/Home/ShowBug">Reset (GET without any query string parameters)</a>

    <hr />

    <form asp-action="@nameof(HomeController.ShowBug)">

        <h4>I expected to see displayed the Model.Name value "Apple", but instead I see the ?Name={value} value "orange" displayed</h4>
        Name: <input asp-for="Name" />  @* asp-for is "binding" to ?Name={value} and NOT Model.Name *@

    </form>
```
13. To see the bug, browse to http://localhost:{port}/Home/ShowBug?Name=orange

For this tag-helper:
    `<input asp-for="Name" />`
I was expecting to get this HTML generated: 
    `<input type="text" id="Name" name="Name" value="Apple" />`
but instead, I get:
    `<input type="text" id="Name" name="Name" value="orange" />`
(note: the value is `"orange"` and not `"Apple"` as expected)

It's as though the view-model's Name property value is being "overriden" by the query-string "Name" parameter value (as per ?name={value})

**Comment by frankabbruzzese** [comment]
@mcrsw ,
It doesn't appears to be a bug, but it appears just the way ModelState work. It is a feature to allow user correct validation errors, by returning the same value he inserted before in a field.
Name value is recorded in ModelState, so if you name something else with "Name" any input field for that property will contain the old "Name" value recorded in ModelState.

ModelState.Clear() in the controller should solve the problem.Once I verify there are no validation errors I aways clear ModelState just to avoid similar problems.

**Comment by mcrsw** [comment]
Hi @frankabbruzzese,

Thank you for your suggestion. I can confirm, that by changing the action method to this:
```
        public IActionResult ShowBug(string name)
        {
            var vm = new ShowBugViewModel
            {
                Name = "Apple",
            };

             ModelState.Clear(); // This "fixes" the problem

            return View(vm);    // See the new view: Home/ShowBug.cshtml
        }
```
the problem is "fixed" - in the sense that
`<input asp-for="Name" />`
does now produce
`<input type="text" id="Name" name="Name" value="Apple" />`

However, this is an HTTP GET action method, so there's no (post back) model validation - as you would expect to happen in an HTTP POST action method.
As such, it does not seem natural to me to involve ModelState.Clear() in my HTTP GET action method. This still feels like a bug to me.

**Comment by dougbu** [comment]
> there's no (post back) model validation

@mcrsw your scenario involves model binding and validation. If it did not, `name` would not have a value. Both model binding and validation update the `ModelState`.

Put another way, model binding handles all action arguments and not just those that get values from a `application/x-www-form-urlencoded` (or `multipart/form-data`) body.

**Comment by frankabbruzzese** [comment]
@mcrsw, 
true that *most of the time* input data submitted by humans come from a post operation.
However, *often* human data comes also from get operations. Typical case is input used to filter data. In this case it is good practice to submit the form with get, to conform to REST semantics and for other practical reasons. So, in general, you might need the same behavior also with get.

For JSon data there are further complications that are far beyond the scope of this discussion. 

Summing up there is no strong reason to break "uniformity" by processing theModelState in a way that depends on the Value provider: breaking uniformity without a good reason would undermine both modularity and maintanaibility of software, often preventing future extensions.

**Comment by mcrsw** [comment]
Hi @dougbu, @frankabbruzzese,

Thank you both for commenting. I believe I may have muddied the waters somewhat - sorry. If I may restate the issue:

I'm still trying to understand why this view code (.cshtml):
```
@model AspNetCoreTagBug.Controllers.ShowBugViewModel

<input asp-for="Name" />
```
renders an input textbox with `value="orange"` when I GET `http://localhost:{port}/Home/ShowBug?Name=orange`

Even though my view model object (as per `@model ShowBugViewModel` at the top of the .cshtml) has it's .Name property hard-coded to "Apple" in the controller's action method (please see code in original post).

I'm sorry for labouring this point, but if the mere presence of the `string name`  in the action signature (and `?name=orange` in the url) can change the behaviour of the `asp-for="Name"`, then that is very unintuitive to me. (And why I still think this behaviour is a bug)

My reference so far has been https://docs.asp.net/en/latest/mvc/views/tag-helpers/intro.html. Yet I cannot see any reference to ModelState on that page. Are there any other docs that might shed some more light on this please?

**Comment by frankabbruzzese** [comment]
Yes the mere presence of "name" in the action method forces the input asp-for="name" to have the the same value as the "name" parameter also if you change the value of "Name" in your ViewModel. Yes it is unintuitive but substantially  the only way to return old values in case user needs to re-edit a submitted form to correctt some errors.

When model binder binds a property it stores bothe possible errors and the original string passed by the client. Now if you don't call ModelState.Clear() and return to the same view that caused the submit the Mvc engine implicitely assumes  that you want the user re-edit its previous input, so it takes the old values from the ModelState instead of what is actually contained in the ViewModel. That's it! So in case you return to the same View and don't want this behavior just call ModelState.Clear()! This is the rule to remeber (I know it is not documented...but the Mvc engine works this way from the beginning ...Mvc2, Mvc3...)

You might object that you don't need this behavior because old values inserted by user are contained in the received ViewModel anyway, so it is enough to return to the View the received ViewModel without modifyng it to enable the user to re-edit its input. Unluckily this assumption is false for the following reasons:

1. Suppose property "name" contains an int and suppose the associated input field is ill formatted because user inserted some not numeric chars. In this case the ViewModel property would contain just 0 because of the failure of the ModelBinding not what the user inserted, since value inserted by the user cannot be stored into an int.
2. Maybe the ViewModel doesn't contain a property for each field submitted, for instance, because the form might be submitted to different action methods depending on some field values and in each case just some of the fields are used by the ViewModel.

**Comment by mcrsw** [comment]
Hi @frankabbruzzese,

Thank you for explaining this to me. Now that I understand a bit more about what's going on under the covers, and after some more precise googling, it has now become clear to me that I have asked a question that's been asked before.
So, if it's any help to anyone else that may come across this in the future, here are some links to the "same" issue, also with explanations:

https://github.com/aspnet/Mvc/issues/5315
https://github.com/aspnet/Mvc/issues/4486

My apologies for this duplicate question, and for erroneously calling this a bug. This issue can now be closed.
